File: Binary Tree/二叉搜索树/98.验证二叉搜索树.cpp
#include <vector>

using namespace std;
/*
 * @lc app=leetcode.cn id=98 lang=cpp
 *
 * [98] 验证二叉搜索树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    void getVector(TreeNode *root, vector<int> &result)
    {
        if (root == nullptr)
        {
            return;
        }

        getVector(root->left, result);
        result.push_back(root->val);
        getVector(root->right, result);
    }
    bool isValidBST(TreeNode *root)
    {
        vector<int> result;
        // 前序遍历二叉树，将tree变成数组
        getVector(root, result);

        //如果数组是有序的，则是二叉搜索树
        for (int i = 1; i < result.size(); i++)
        {
            if (result[i - 1] >= result[i])
            {
                return false;
            }
        }

        return true;
    }
};
// @lc code=end

File: Binary Tree/二叉搜索树/98.验证二叉搜索树.ts
/*
 * @lc app=leetcode.cn id=98 lang=typescript
 *
 * [98] 验证二叉搜索树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isValidBST(root: TreeNode | null): boolean {
  // 中序遍历，将二叉搜索树转换为数组
  const array = []
  const getArray = (root) => {
    if (!root) {
      return
    }
    // 使用中序遍历是为了符合二叉搜索树的特性
    // 遍历完他就是一个升序数组
    getArray(root.left)
    array.push(root.val)
    getArray(root.right)
  }

  getArray(root)
  // 如果不是升序数组，则不是二叉搜索树
  for (let i = 1; i < array.length; i++) {
    if (array[i] <= array[i - 1]) {
      return false
    }
  }

  return true
};
// @lc code=end

